sample_annotator/ontology/triad_normalizer.py

```diff
@@ -0,0 +1,125 @@
+envo_label_frame = envo_label_frame[['id', 'lbl']]
+# convert IRIs <-> CURIEs
+envo_label_frame['curie'] = envo_label_frame['id'].apply(curie_mapper.getCurie)
+envo_label_frame['tidied'] = envo_label_frame['lbl'].apply(standard_tidy)
+
+
+def get_bio_registry_content(url="https://raw.githubusercontent.com/bioregistry/bioregistry/main/src/bioregistry/data"
+                                 "/bioregistry.json") -> dict:
+    br_dict = get_web_json(url)
+    return br_dict
+
+
+# this should be global ie not re-downloaded and recreated for each normalization
+bio_registry_content = get_bio_registry_content()
+
+
+def get_pattern_from_bioregistry(br_dict, prefix: str, authority="miriam") -> str:
+    # convert prefix and authority to lowercase?
+    # some authorities have patterns and others don't
+    # some may be more specific than others
+    # the pattern will probably be bookended by ^ and $
+    # will probably want to remove those
+    # could these tests be more compact?
+    # what will be returned in any of the fall-through cases?
+    if isinstance(br_dict, dict):
+        if prefix in br_dict.keys():
+            if isinstance(br_dict[prefix], dict) and authority in br_dict[prefix].keys():
+                if isinstance(br_dict[prefix][authority], dict) and 'pattern' in br_dict[prefix][authority].keys():
+                    pattern = br_dict[prefix][authority]['pattern']
+                    pattern = re.sub(r'^\^', '', pattern)
+                    pattern = re.sub(r'\$$', '', pattern)
+                    return pattern
+
+
+envo_patttern = get_pattern_from_bioregistry(bio_registry_content, "envo")
+
+
+def extract_cuire(input_str, term_pattern=envo_patttern) -> dict:
+    term_search = re.search(term_pattern, input_str)
+    # include or exclude brackets from match?
+    # we have seen cases where envo was misspelled
+    # or there was a alpha local part
+    # or ...
+    if term_search:
+        # possibility of multiple matches?
+        term_match = term_search.group(0)
+        # print(term_match)
+        depleted = re.sub(term_match, '', input_str)
+        depleted = standard_tidy(depleted)
+        asserted_label = envo_label_frame.loc[envo_label_frame['curie'] == term_match, "lbl"]
+        # print(asserted_label)
+        if len(asserted_label) == 1:
+            al_val = asserted_label.squeeze()
+            print(al_val)
+            return {"depleted": depleted, "term_match": term_match, "asserted_label": al_val}
+        else:
+            return {"depleted": depleted, "term_match": term_match}
+    else:
+        tidied = standard_tidy(input_str)
+        return {"tidied": tidied}
+
+
+# returned pattern should contain
+# one or more pipe-delimited
+# OBO foundry label and [term id] (usually EnvO)
+# "terrestrial biome [ENVO:00000446]|marine biome [ENVO:00000447]"
+def normalize_triad_slot(raw_triad_value: str) -> str:
+    # start by looking for exact matches with rdftab?
```

Comment:
consider docstrings

Reply:
moved a lot of the comments to docstrings. Probably still room for improvement, like following somebody's docstring format